### DISCUSSION TITLE
pyright equivalent of strictNullChecks

### DISCUSSION THREAD

**Discussion by mehdigmira** [microsoft/pyright#6016] [Q&A]
Hey there,

Is there a Pyright equivalent of Typescript's "strictNullChecks" ?

Looking at the documentation, I find related parameters: 
- strictParameterNoneValue 
- reportOptionalSubscript
- reportOptionalMemberAccess
- reportOptionalCall 
- reportOptionalIterable 
- reportOptionalContextManager 
- reportOptionalOperand 

But they don't exactly behave the same. In Typescript when strictNullChecks is disabled, inferred types are not optional, so the following code logs no error:

```typescript
declare const loggedInUsername: string;

const users = [
  { name: 'Oby', age: 12 },
  { name: 'Heera', age: 32 }
];

const loggedInUser = users.find(u => u.name === loggedInUsername);  // when strictNullChecks=false this inferred as { name: string; age: number }
export const f = (user: { name: string; age: number }) => user;
f(loggedInUser) // when strictNullChecks=false, no error is reported
```

But in Python:

```python
# pyright: strictParameterNoneValue=false, reportOptionalSubscript=false, reportOptionalMemberAccess=false, reportOptionalCall =false, reportOptionalIterable =false, reportOptionalContextManager =false, reportOptionalOperand =false

from __future__ import annotations

d = {"a": 1, "b": 2}

def f(x: int):
    ...

y = d.get("c")

f(y)  #  reportGeneralTypeIssues reported here
```

I ask this because in the case of non production code (such as tests or scripts), it's common to not worry about optional values, since we know that the result is never None at runtime. So when opening up such files with pylance you end up a lot of reportGeneralTypeIssues being issued.

Maybe there is a way to achieve such a thing ?

**Comment by erictraut** [comment]
In your code sample, the `reportGeneralTypeIssues` error comes from the fact that you're attempting to assign a value of type `int | None` to a parameter of type `str`. I suspect you meant to annotate the parameter as `int`.

If I understand you correctly, you're looking for something like `reportOptionalAssignment`. If set to false, this option would tell the type checker to ignore cases where a value of type `T | None` was assigned to a variable or parameter of type `T`. 

I think such a feature would be feasible to implement, but it would be quite a bit of work to implement because it would affect quite a few places in the type checking logic. I'd also need to think about the implications for TypeVar constraint solving and type narrowing on assignment. 

It's a feature that we could consider adding in the future if there's sufficient interest (e.g. upvotes).

**Reply by mehdigmira** [reply]
Sorry, edited the code sample.

yes `reportOptionalAssignment` as you describe it would solve the issue !

The way Typescript implements this seems to be that the inferred type of variables is never optional. Just out of curiosity, would adding a step that removes "| None" from inferred types be a viable way to do this ?

**Comment by aspizu** [comment]
May I suggest the use of either
```py
x: T | None
assert x is not None
f(x)
```
or
```py
x: T | None
f(cast(T, x))
```
or
```py
def is_some(x: T | None) -> T:
   assert x is not None
   return x
x: T | None
f(is_some(x))
```
